libenv/sysinfo.c

```diff
@@ -1129,7 +1129,14 @@ static void OSClasses(EvalContext *ctx)
     {
         Xen_Domain(ctx);
     }
-
+#ifdef XEN_CPUID_SUPPORT
+    else if (Xen_Hv_Check())
```

Comment:
If /proc/xen/capabilities exists calling the function Xen_Domain() will set the class 'xen' and also determine if its dom0 or domU so there is no need to call Xen_Hv_Check because we already set the classes that need to be set.

If that file doesn't exist (xenfs is not loaded or the guest being virtualized doesn't support it for example) you can call cpuid and check if the vendor id is 'XenVMMXenVMM' which will work on any x86 host.

However, right now this function gets compiled out unless its being built on Linux but it addresses the issue documented in CFE-1563. I plan on refactoring this code a little bit so that at the very least it can also get called if the platform is FreeBSD to address CFE-2203.

(I updated the sysutils/cfengine39 port on FreeBSD to fix this with a patch when the port but since I only need to worry about FreeBSD when that port is being built its a bit shoe-horned in. If you want to take a look at that change the code review is at https://reviews.freebsd.org/D8607)